File: executor-ts/src/utils/sanitize.ts
import path from "path";

/**
 * Resolves `userPath` relative to `baseDir`, then verifies the resolved
 * absolute path stays within `baseDir`. Throws on path traversal attempts.
 */
export function sanitizePath(userPath: string, baseDir: string): string {
  const resolved = path.resolve(baseDir, userPath);
  const normalizedBase = path.resolve(baseDir) + path.sep;
  if (!resolved.startsWith(normalizedBase) && resolved !== path.resolve(baseDir)) {
    throw new Error(
      `Path traversal blocked: "${userPath}" resolves outside "${baseDir}"`
    );
  }
  return resolved;
}

File: btc-treasury/internal/utils/sanitize.go
package utils

import (
	"fmt"
	"path/filepath"
	"strings"
)

// SanitizePath resolves raw path relative to base and ensures the resolved path
// stays within base. Returns the resolved clean path or an error if path traversal is detected.
func SanitizePath(raw string, base string) (string, error) {
	absBase, err := filepath.Abs(base)
	if err != nil {
		return "", fmt.Errorf("cannot resolve base directory: %w", err)
	}

	resolved := filepath.Join(absBase, raw)
	absResolved, err := filepath.Abs(resolved)
	if err != nil {
		return "", fmt.Errorf("cannot resolve path %q: %w", raw, err)
	}

	cleanBase := filepath.Clean(absBase)
	cleanResolved := filepath.Clean(absResolved)

	// Ensure cleanResolved is inside cleanBase.
	// Since cleanBase is clean, we can check if cleanResolved has cleanBase as prefix.
	// We append a path separator to ensure we match whole directory components.
	sep := string(filepath.Separator)
	prefix := cleanBase
	if !strings.HasSuffix(prefix, sep) {
		prefix += sep
	}

	if cleanResolved != cleanBase && !strings.HasPrefix(cleanResolved, prefix) {
		return "", fmt.Errorf("path traversal blocked: %q escapes %q", raw, base)
	}

	return cleanResolved, nil
}
